mapeditor/inspector/townevent.cpp

```diff
@@ -0,0 +1,320 @@
+/*
+ * townevent.cpp, part of VCMI engine
+ *
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+
+#include "../StdInc.h"
+#include "townbuildingswidget.h"
+#include "townevent.h"
+#include "ui_townevent.h"
+#include "../../lib/constants/NumericConstants.h"
+#include "../../lib/constants/StringConstants.h"
+
+TownEvent::TownEvent(CGTownInstance & t, QListWidgetItem * item, QWidget * parent) :
+	QDialog(parent),
+	ui(new Ui::TownEvent),
+	town(t),
+	item(item)
+{
+	ui->setupUi(this);
+
+	ui->buildingsTree->setModel(&buildingsModel);
+
+	params = item->data(Qt::UserRole).toMap();
+	ui->eventFirstOccurrence->setMinimum(1);
+	ui->eventFirstOccurrence->setMaximum(999);
+	ui->eventRepeatAfter->setMaximum(999);
+	ui->eventNameText->setText(params.value("name").toString());
+	ui->eventMessageText->setPlainText(params.value("message").toString());
+	ui->eventAffectsCpu->setChecked(params.value("computerAffected").toBool());
+	ui->eventAffectsHuman->setChecked(params.value("humanAffected").toBool());
+	ui->eventFirstOccurrence->setValue(params.value("firstOccurrence").toInt()+1);
+	ui->eventRepeatAfter->setValue(params.value("nextOccurrence").toInt());
+
+	initPlayers();
+	initResources();
+	initBuildings();
+	initCreatures();
+
+	show();
+}
+
+TownEvent::~TownEvent()
+{
+	delete ui;
+}
+
+void TownEvent::initPlayers()
+{
+	for (int i = 0; i < PlayerColor::PLAYER_LIMIT_I; ++i)
+	{
+		bool isAffected = (1 << i) & params.value("players").toInt();
```

Comment:
what is this bit magic? can't you store an array?

Reply:
Code copied from timedevent.cpp, looks like players in events are stored in bit field
https://github.com/vcmi/vcmi/blob/3c611ffa5b76686156d89aded68e87fc4e34c2b6/lib/mapping/CMapDefines.h#L39

Reply:
Might be a good idea to replace with std::set<PlayerColor>, but not insisting.

Reply:
ah, since it's coming not from editor settings, then it's fine of course